### PR TITLE
BUG: failure to cast all-int floating dtypes when setting into int dtypes

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [x] whatsnew entry

Overlap with (but not rebased on top of) #44275

### CODE REVIEW

**Review by jreback — APPROVED** [review]
lgtm @phofl i fyou can have a look

**Comment by jreback** [comment]
user facing at all?

**Comment by jbrockmendel** [comment]
> user facing at all?

yes, just updated with whatsnew

File: pandas/core/series.py [hunk 1]
```diff
@@ -1096,9 +1096,26 @@ def __setitem__(self, key, value) -> None:
             if com.is_bool_indexer(key):
                 key = check_bool_indexer(self.index, key)
                 key = np.asarray(key, dtype=bool)
+
```
Comment (phofl) [inline pandas/core/series.py:1099]: This is from #44275?

Comment (jbrockmendel) [reply pandas/core/series.py:1099]: yes (mentioned in OP)

Comment (phofl) [reply pandas/core/series.py]: Yeah, thx. Was a bit confused why I could not see that master was updated here

File: pandas/tests/indexing/multiindex/test_setitem.py [hunk 2]
```diff
@@ -196,16 +196,29 @@ def test_multiindex_assignment(self):
...
+            dtype=np.int64,
         )
 
+        # arr can be losslessly cast to int, so this setitem is inplace
         df.loc[4, "c"] = arr
-        exp = Series(arr, index=[8, 10], name="c", dtype="float64")
-        tm.assert_series_equal(df.loc[4, "c"], exp)
+        exp = Series(arr, index=[8, 10], name="c", dtype="int64")
+        result = df.loc[4, "c"]
+        tm.assert_series_equal(result, exp)
```
Comment (phofl) [inline pandas/tests/indexing/multiindex/test_setitem.py:215]: Should we use a view here to show, that the underlying object changed?

**Comment by phofl** [comment]
small comment, otherwise lgtm

File: pandas/tests/indexing/multiindex/test_setitem.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/tests/indexing/multiindex/test_setitem.py:215]: good call, turns out the ArrayManager subs in a new object

Comment (jbrockmendel) [reply pandas/tests/indexing/multiindex/test_setitem.py:216]: updated+green

**Review by phofl — APPROVED** [review]
Lgtm

**Comment by phofl** [comment]
Thx @jbrockmendel

**Pull request merged**